designs/disruption-controls-by-reason.md

```diff
@@ -0,0 +1,312 @@
+- drifted = 10 - 3: Disruption allowed as drifted bucket isn't full.
+- underutilized = 15 - 6: Disruption allowed since 15 - 6 > 0.
+- Default = 5 - expired - empty = 0: No disruption allowed for reasons other than consolidation and drift.
+
+**Second Equation**
+- Total Disrupting = 3 + 6 + 5 = 14.
+- drifted = 10 - 14: No disruption allowed due to drift.
+- underutilized = 15 - 14: Disruption allowed for at least one node due to consolidation.
+- Default: 5 - 14: No disruption allowed for other methods.
+
+#### Decision
+The second equation simplifies reasoning for cluster operators. A node can be tagged for multiple disruption reasons (e.g., empty, drifted, underutilized), making it challenging to precisely calculate in-flight disruptions. The second equation streamlines this process by providing a clearer view of the overall disruption impact.
+### Q: How should we handle an unspecified default reason? 
+```yaml
+budgets: 
+  - nodes: 10
+    reasons: [drift, underutilized]
+    schedule: "* * * * *"
+  - nodes: 5 
+    reasons: [empty] 
+    schedule: "* * * * *" 
+```
+In the case of a budget like above, default is undefined. Should karpenter assume the user doesn't want to disrupt any other reasons? Or should we assume that if a default is unspecified, they want us to disrupt anyway?  
+The intuitive options if there is no active default budget is to allow disruption of either 0 or total number of nodes(meaning unbounded disruption).
+Lets choose total number of nodes, since this allows the user to also specify periods where no nodes are to be disrupted of a particular type of disruption, and makes more sense with the existing karpenter behavior today.
+
+# API Design
+## Approach A: Extending the Budget API to specify a reason 
+This section contrasts two approaches for specifying disruption reasons in the v1beta1 nodepool API: a list of reasons (List Approach) and a single reason per budget (Single Reason Approach).
+
+### List Approach: Multiple Reasons per Budget
+This approach allows specifying multiple disruption methods within a single budget entry. It is proposed to add a field Reasons to the budgets, which can include a list of reasons this budget applies to.
+#### Proposed Spec
+Add a simple field "reasons" is proposed to be added to the budgets. 
+```go
+// Budget defines when Karpenter will restrict the
+// number of Node Claims that can be terminating simultaneously.
+type Budget struct {
+      // Reasons is a list of methods for disruption that apply to this budget. If Reasons is not set, this budget applies to all methods.
+      // If a reason == "default", it will apply to all reasons that don't have an active budget. If a reason is set, it will only apply to that method. If multiple reasons are specified,
+      // this budget will apply to all of them. If a budget is not specified for a method, the default budget will be used.
+      // allowed reasons are "default", "consolidation", "expired", "empty", "drift"
```

Comment:
I think we discussed that default == omitting this field, right? Can we update this description to match that? Looks like the below kubebuilder validation reflects this. 
```suggestion
      // If a reason is set, it will only apply to that method. If multiple reasons are specified,
      // this budget will apply to all of them. If a budget is not specified for a method, the default budget will be used.
      // allowed reasons are "consolidation", "expired", "empty", "drift"
```

Reply:
Also what does a "default budget" mean here? Can we just say it's unbounded?